client/src/main/java/org/apache/uniffle/client/impl/ShuffleWriteClientImpl.java

```diff
@@ -123,27 +125,71 @@ public ShuffleWriteClientImpl(
     this.replicaWrite = replicaWrite;
     this.replicaRead = replicaRead;
     this.replicaSkipEnabled = replicaSkipEnabled;
-    this.dataTransferPool = new ForkJoinPool(dataTranferPoolSize);
+    this.dataTransferPool = Executors.newFixedThreadPool(dataTranferPoolSize);
     this.dataCommitPoolSize = dataCommitPoolSize;
     this.unregisterThreadPoolSize = unregisterThreadPoolSize;
     this.unregisterRequestTimeSec = unregisterRequestTimeSec;
   }
 
+  @VisibleForTesting
+  public static boolean waitUntilDoneOrFail(List<Future<Boolean>> futures, boolean allowFastFail) {
+    int expected = futures.size();
+    int failed = 0;
+
+    List<Future> finished = new ArrayList<>();
+    while (true) {
+      for (Future<Boolean> future : futures) {
+        if (future.isDone() && !finished.contains(future)) {
+          finished.add(future);
+          try {
+            if (!future.get()) {
+              failed++;
+            }
+          } catch (Exception e) {
+            failed++;
+          }
+        }
+      }
+
+      if (expected == finished.size()) {
+        return failed <= 0;
+      }
+
+      if (failed > 0 && allowFastFail) {
+        futures.stream().filter(x -> !x.isDone()).forEach(x -> x.cancel(true));
+        return false;
+      }
+
+      Uninterruptibles.sleepUninterruptibly(10, TimeUnit.MILLISECONDS);
+    }
+  }
+
   private boolean sendShuffleDataAsync(
       String appId,
       Map<ShuffleServerInfo, Map<Integer, Map<Integer, List<ShuffleBlockInfo>>>> serverToBlocks,
       Map<ShuffleServerInfo, List<Long>> serverToBlockIds,
-      Map<Long, AtomicInteger> blockIdsTracker) {
+      Map<Long, AtomicInteger> blockIdsTracker, boolean allowFastFail,
+      Supplier<Boolean> needCancelRequest) {
+
+    if (serverToBlockIds == null) {
+      return true;
+    }
+
     // If one or more servers is failed, the sending is not totally successful.
-    AtomicBoolean isAllServersSuccess = new AtomicBoolean(true);
-    if (serverToBlocks != null) {
-      dataTransferPool.submit(() -> serverToBlocks.entrySet().parallelStream().forEach(entry -> {
+    List<Future<Boolean>> futures = new ArrayList<>();
+    for (Map.Entry<ShuffleServerInfo, Map<Integer, Map<Integer, List<ShuffleBlockInfo>>>> entry :
+        serverToBlocks.entrySet()) {
+      Future<Boolean> future = dataTransferPool.submit(() -> {
+        if (!needCancelRequest.get()) {
```

Comment:
When `needCancelRequest` is true, we should return true.

Reply:
Done